README.md
# taskmaster

Client / server based project.

Client can start / stop / reload multiple programs with request to the server and the server can monitor them
based on the configuration file.

The server can relaunch program which were interrupt from a signal that wasn't in the configuration file


history.txt
start cat
stop cat
start cat
stop cat
stop ca
shutdown
start cat
status
shutdown
start cat
status
shutdown
start cat
status
l
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
shutdown
start ls
stop all
status
shutdown
start ls
start cat
shutdown
stop all


client.py
import time
import socket
import configurationClass as conf

class Client:

    host = str()
    port = int()
    history = list()
    conf_file = conf.ConfigurationFile()

    def __init__(self, sock=None):

        if sock == None:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        else:
            self.sock = sock

    def clientConnexion(self, host='127.0.0.1', port=12800):

        self.host = host
        self.port = port

        try:
            self.sock.connect((host, port))
            return 0
        except:
            print("Error: Can't connect to taskmaster daemon")
            return 1

    def sendMessage(self, message):

        try:
            message = bytes(message, 'UTF-8')
            self.sock.send(message)
            return 0
        except:
            print("\nError: Can't delivered message to taskmasterd")
            return 1

    def receiveMessage(self, buff=1024):
        self.sock.settimeout(0.05)
        try:
            message = self.sock.recv(buff)
            return message.decode()
        except socket.timeout:
            return ''
        except:
            print("Error: can't receive message from server")

    def closeConnection(self):

        self.sock.close()
        print("Connection close. Bye bye 3:)")

    def historyFileWrite(self, history):

        history_path = self.conf_file.parsed_file['taskmasterctl'].get('history_file', './history.txt')
        try:
            with open(history_path, 'w+') as history_file:
                for command in history:
                    history_file.write(command + '\n')
        except:
            print("Error: Can't open history file")
            return []
        history_file.close()

    def historyFileLoad(self):

        history_path = self.conf_file.parsed_file['taskmasterctl'].get('history_file', './history.txt')
        try:
            with open(history_path, 'r') as history_file:
                for current_line in history_file.readlines():
                    self.history.append(current_line.rstrip())
        except:
            return []
        history_file.close()
        return self.history

    def parse(self, line):

        if not line:
            return 2
        line = line.split()
        for arg in line:
            arg.strip()
        cmd = line[0]
        if cmd != 'start' and cmd != 'restart' and cmd != 'stop' and cmd != 'status' and cmd != 'reload' and cmd != 'close' and cmd != 'help' and cmd != 'shutdown':
            return 1
        if cmd == 'help':
            return 3
        return (' '.join(line))


daemonClass.py
import server
import threading
import os

class createDaemon:

    UMASK = 0
    WORKDIR = "/Users/robechon/Taskmaster"
    MAXFD = 1024
    
    if(hasattr(os, "devnull")):
        REDIRECT_TO = os.devnull
    else:
        REDIRECT_TO = "/dev/null"

    def __init__(self):
        
        try:
            pid = os.fork()
        except OSError, e:
            raise Exception, "%s [%d]" % (e.strerror, e.errno)
       
        if pid == 0:
            os.setsid()
            try:
                pid = os.fork
            except OSError, e:
                raise Exception, "%s [%d]" % (e.strerror, e.errno)
            if pid == 0:   
                os.chdir(WORKDIR)
                os.umask(UMASK)
            else:
                os._exit(0)
        else:
            os._exit(0)


if __name__ == '__main__':

    retCode = createDaemon()
    while True:
        server.Server('127.0.0.1', 12800).listen()


init_termios.py
import time
import curses, signal, os
import client
import configurationClass as conf
from curses import textpad

class InitTermios:

    configuration = conf.ConfigurationFile()
    client_created = client.Client()
    prompt = configuration.parsed_file['taskmasterctl']['prompt'] + " > "
    cursor_x = 0
    cursor_y = 0
    window_x = 0
    window_y = 0
    command_x = 0
    command_y = 9
    output_x = 0
    output_y = 15
    stdscr = 0
    line = []
    cant_print = False
    history = []
    h = 0
    i = 0

    def __init__(self):
        if self.client_created.clientConnexion() == 1:
            return
        self.history = self.client_created.historyFileLoad()
        self.h = len(self.history)
        signal.signal(signal.SIGINT, self.catchSigInt)
        signal.signal(signal.SIGWINCH, self.catchSigWinch)
        curses.wrapper(self.initWindow)

    def catchSigInt(self, signal, frame):
        self.stdscr.erase()
        self.tryDisplay()

    ##
    # A REVOIR EN FONCTION DES RETOUR POSSIBLE
    ##
    def catchSigWinch(self, signal, frame):
        rows, columns = os.popen('stty size', 'r').read().split()
        # self.stdscr.resize(int(rows), int(columns))
        curses.resizeterm(int(rows), int(columns))
        (get_window_y, get_window_x) = self.stdscr.getmaxyx()
        self.window_x = get_window_x
        self.window_y = get_window_y
        self.stdscr.erase()
        self.tryDisplay()
        # (cursor_tuple_y, cursor_tuple_x) = self.stdscr.getyx()
        # print(str(self.window_x), str(self.window_y), str(columns), str(rows))

    def printInInput(self):
        len_max_command = self.window_x * (self.output_y - self.command_y) - len(self.prompt) - (self.output_y - self.command_y) - 1
        if (len(self.line) > len_max_command):
            self.stdscr.insnstr(self.cursor_y, self.cursor_x, "Can't display input", 0)
        elif len(self.line) > 0:
            for line_char in self.line:
                self.stdscr.insch(self.cursor_y, self.cursor_x, line_char)
                self.cursor_x += 1
                if self.cursor_x == self.window_x - 1:
                    self.cursor_x = 0
                    self.cursor_y += 1
                self.stdscr.move(self.cursor_y, self.cursor_x)
                self.i += 1

    def printInOutput(self, message):
        output_max_len = (self.window_x - 2) * (self.window_y - self.output_y - 2) - (self.window_y - self.output_y - 2)
        end_line = 0
        for letter in message:
            if letter == '\n':
                end_line += 1
        # end_line = message.count('\n')
        # end_line += 1
        if end_line == 0:
            end_line = 1
        #
        # A VOIR POUR LE X MAX
        #
        if (len(message) > output_max_len) or end_line > (self.window_y - self.output_y - 2):
            self.stdscr.insnstr(self.output_y + 1, 2, "Can't display output", 0)
        else:
            message_split = message.split('\n')
            put_y = self.output_y
            for splitted in message_split:
                put_x = 2
                for letter in splitted:
                    self.stdscr.insch(put_y + 1, put_x, letter)
                    put_x += 1
                    if put_x >= self.window_x - 2:
                        put_x = 2
                        put_y += 1
                        if put_y - self.output_y - 1 >= (self.window_y - self.output_y - 2):
                            self.stdscr.erase()
                            self.resetWindow()
                            self.stdscr.insnstr(self.output_y + 1, 2, "Can't display output", 0)
                put_y += 1
            self.createFrame()


    def displayHeader(self):
        header1 = "  _____         _    __  __           _"
        header2 = " |_   _|_ _ ___| | _|  \/  | __ _ ___| |_ ___ _ __"
        header3 = "   | |/ _` / __| |/ / |\/| |/ _` / __| __/ _ \ '__|"
        header4 = "   | | (_| \__ \   <| |  | | (_| \__ \ ||  __/ |"
        header5 = "   |_|\__,_|___/_|\_\_|  |_|\__,_|___/\__\___|_|"

        center_header_x = self.window_x - 51
        self.stdscr.insnstr(0, int(center_header_x / 2), header1, 0)
        self.stdscr.insnstr(1, int(center_header_x / 2), header2, 0)
        self.stdscr.insnstr(2, int(center_header_x / 2), header3, 0)
        self.stdscr.insnstr(3, int(center_header_x / 2), header4, 0)
        self.stdscr.insnstr(4, int(center_header_x / 2), header5, 0)
        connexion_data = "Connexion established with {} on port {}" .format(self.client_created.host, self.client_created.port)
        center_connexion_data_x = self.window_x - len(connexion_data)
        self.stdscr.insnstr(self.command_y - 2, int(center_connexion_data_x / 2), connexion_data, 0)

    def resetWindow(self):
        self.i = 0
        self.cursor_y = self.command_y
        self.cursor_x = self.command_x + len(self.prompt)
        self.displayHeader()
        self.createFrame()
        self.stdscr.insnstr(self.cursor_y, 0, self.prompt, 0)
        self.stdscr.move(self.cursor_y, self.cursor_x)
        self.printInInput()
        self.cant_print = False

    def tryDisplay(self):
        if self.window_x < 53 or self.window_y < 20:
            self.stdscr.erase()
            self.stdscr.insnstr(0, 0, "Can't display Taskmaster", 0)
            self.cant_print = True
        # (cursor_tuple_y, cursor_tuple_x) = self.stdscr.getyx()
        else:
            self.resetWindow()

    def input(self, stdscr, close_window):
        end_read = False
        # if self.cant_print == True:
        #     end_read = True
        while (end_read == False):
            response = self.client_created.receiveMessage()
            if response != '' and response != '\n' and response != None:
                self.printInOutput(response)
                self.cursor_y = self.command_y
                self.cursor_x = self.command_x + len(self.prompt)
                self.stdscr.move(self.cursor_y, self.cursor_x)
            key = stdscr.getch()
            if self.cant_print == False:
                len_max_command = self.window_x * (self.output_y - self.command_y) - len(self.prompt) - (self.output_y - self.command_y) - 1
                #Return
                if (key == 10):
                    end_read = True
                    stdscr.erase()
                    # self.cursor_y += 1
                    result = self.client_created.parse(''.join(self.line))
                    if len(self.line) > 0:
                        self.h = len(self.history)
                        if self.h == 0:
                            self.history.append(''.join(self.line))
                            self.h += 1
                        elif self.history[-1] != ''.join(self.line):
                            self.history.append(''.join(self.line))
                            self.h += 1
                    if result == 1:
                        self.printInOutput("Error: bad command")
                    elif result == 2:
                        continue
                    elif result == 3:
                        self.printInOutput("""COMMANDS: - start <program>
                - restart <program>
                - stop <program>
                - status <program>
                - reload
                - close""")
                    else:
                        self.client_created.sendMessage(result)
                    if (''.join(self.line) == "close"):
                        close_window = True
                    self.line = []
                #Ctrl-C / Check Twice
                # elif (key == -1):
                #     end_read = True
                #     self.line = []
                #     stdscr.erase()
                    # self.cursor_y += 1
                # #Ctrl-L / Clear
                # elif key == 12:
                #     stdscr.erase()
                #     self.cursor_y = self.command_y
                #     stdscr.move(self.cursor_y, self.cursor_x)
                #     end_read = True
                #Esc / Ctrl-D
                elif key == 27 or key == 4:
                    end_read = True
                    close_window = True
                    self.client_created.sendMessage('close')
                #Suppr
                elif key == 127 and self.i > 0:
                    self.cursor_x -= 1
                    if self.cursor_x < 0:
                        self.cursor_x = self.window_x - 2
                        self.cursor_y -= 1
                    stdscr.delch(self.cursor_y, self.cursor_x)
                    del self.line[self.i - 1]
                    self.i -= 1
                #Delete
                elif key == 330 and self.i < len(self.line):
                    stdscr.delch(self.cursor_y, self.cursor_x)
                    del self.line[self.i]
                #Home
                elif key == 262:
                    self.i = 0
                    self.cursor_x = self.command_x + len(self.prompt)
                    self.cursor_y = self.command_y
                    stdscr.move(self.command_y, self.cursor_x)
                #End
                elif key == 360:
                    self.i = len(self.line)
                    self.cursor_x = int((len(self.line) + len(self.prompt)) % self.window_x) + self.command_x + 1
                    self.cursor_y = int((len(self.line) + len(self.prompt)) / self.window_x) + self.command_y
                    stdscr.move(self.cursor_y, self.cursor_x)
                #Arrow_up
                elif key == 259 and self.h > 0:
                    self.h -= 1
                    self.line = []
                    for j in self.history[self.h]:
                        self.line.append(j)
                    self.stdscr.erase()
                    self.resetWindow()
                #Arrow_down
                elif key == 258 and self.h < len(self.history):
                    if self.h < len(self.history) - 1:
                        self.h += 1
                        self.line = []
                        for j in self.history[self.h]:
                            self.line.append(j)
                        self.stdscr.erase()
                        self.resetWindow()
                    else:
                        self.h += 1
                        self.line = []
                        self.stdscr.erase()
                        self.resetWindow()
                #Arrow_left
                elif key == 260 and self.cursor_x > len(self.prompt):
                    self.cursor_x -= 1
                    if self.cursor_x < 0:
                        self.cursor_x = self.window_x - 2
                        self.cursor_y -= 1
                    self.i -= 1
                    stdscr.move(self.cursor_y, self.cursor_x)
                #Arrow_right
                elif key == 261 and self.i < len(self.line):
                    self.cursor_x += 1
                    if self.cursor_x == self.window_x - 1:
                        self.cursor_x = 0
                        self.cursor_y += 1
                    self.i += 1
                    stdscr.move(self.cursor_y, self.cursor_x)
                #Printable
                elif key >= 32 and key < 127 and len(self.line) < len_max_command:
                    stdscr.insch(self.cursor_y, self.cursor_x, key)
                    self.cursor_x += 1
                    if self.cursor_x == self.window_x - 1:
                        self.cursor_x = 0
                        self.cursor_y += 1
                    stdscr.move(self.cursor_y, self.cursor_x)
                    self.line.insert(self.i, chr(key))
                    self.i += 1
        if (close_window == True):
            return True
        else:
            return False

    def windowManager(self, stdscr):
        close_window = False
        while (close_window == False):
            self.tryDisplay()
            close_window = self.input(stdscr, close_window)
            # if len(self.line) > 0:
            #     print("  ##### ", ''.join(self.line))
            if (close_window == True):
                self.client_created.closeConnection()
        if len(self.history) >= 100:
            self.history = self.history[60:]
        self.client_created.historyFileWrite(self.history)

    def initWindow(self, stdscr):
        # Clear screen
        stdscr.clear()
        self.stdscr = stdscr
        self.stdscr.nodelay(True)
        # (cursor_tuple_y, cursor_tuple_x) = stdscr.getyx()
        (self.window_y, self.window_x) = self.stdscr.getmaxyx()
        self.windowManager(stdscr)
        stdscr.refresh()

    def createFrame(self):
        textpad.rectangle(self.stdscr, self.output_y, 0, self.window_y - 1, self.window_x - 2)


server.py
import socket
import threading
import sys
import time
import os
import signal
import configurationClass as Conf
from threading import Thread

class Server:

    host = str()
    port = int()
    program_dict = {}
    arg_all = False
    current_client = 0
    conf = Conf.ConfigurationFile()

    def __init__(self, host, port):

        self.host = host
        self.port = port
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.sock.bind((host, port))
        for program in self.conf.parsed_file['programs']:
            self.program_dict[program] = ['STOPPED']

    def start(self, program):
        self.execute(self.conf.parsed_file['programs'], program)
        return

    def stop(self, program):
        if self.program_dict[program][0] == 'RUNNING':
            #
            # SLEEP EST POUR LORSQUE LE PROGRAMME NE S ARRETE TOUJOURS PAS
            #
            time.sleep(float(self.conf.parsed_file['programs'][program].get('stopwaitsecs', 10)))
            stopsignal = self.conf.parsed_file['programs'][program].get('stopsignal', "TERM")
            if stopsignal == "TERM":
                os.kill(self.program_dict[program][1], signal.SIGTERM)
            elif stopsignal == "HUP":
                os.kill(self.program_dict[program][1], signal.SIGHUP)
            elif stopsignal == "INT":
                os.kill(self.program_dict[program][1], signal.SIGINT)
            elif stopsignal == "QUIT":
                os.kill(self.program_dict[program][1], signal.SIGQUIT)
            elif stopsignal == "KILL":
                os.kill(self.program_dict[program][1], signal.SIGKILL)
            elif stopsignal == "USR1":
                os.kill(self.program_dict[program][1], signal.SIGUS1)
            elif stopsignal == "USR2":
                os.kill(self.program_dict[program][1], signal.SIGUSR2)
            self.current_client.send("{}: stopped".format(program).encode('utf-8'))
            print("{}: stopped".format(program))
        else:
            if self.arg_all == False:
                self.current_client.send("{}: ERROR (not running)".format(program).encode('utf-8'))
                print("{}: ERROR (not running)".format(program))
        return

    def restart(self, message_list):
        self.stop(message_list)
        self.start(message_list)

    def printStatus(self, program):
        if len(self.program_dict[program]) == 1:
            return ("{}       {}".format(program, self.program_dict[program][0]))
            print("{}       {}".format(program, self.program_dict[program][0]))
        else:
            return ("{}       {}  pid {}".format(program, self.program_dict[program][0], self.program_dict[program][1]))
            print("{}       {}  pid {}".format(program, self.program_dict[program][0], self.program_dict[program][1]))

    def status(self, program):
        response = []
        if program == '':
            for key in self.program_dict.keys():
                response.append(self.printStatus(key))
            self.current_client.send('\n'.join(response).encode('utf-8'))
        else:
            self.current_client.send(self.printStatus(program).encode('utf-8'))

    def multipleAction(self, message_list, action):

        if (len(message_list) == 0):
            return
        else:
            if message_list[0] in self.program_dict.keys():
                multi_process = Thread(target=action, args=(message_list[0],))
                multi_process.start()
            else:
                self.current_client.send("{}: ERROR (no such process)".format(message_list[0]).encode('utf-8'))
                print("{}: ERROR (no such process)".format(message_list[0]))
            del message_list[0]
            self.multipleAction(message_list, action)

    def redirectMessage(self, message):

        message_list = message.split(" ")
        if (len(message_list) == 1):
            if (message_list[0] == 'status'):
                self.status('')
                return
            elif(message_list[0] == 'reload'):
                # self.reload()
                return
        else:
            if (message_list[0] == 'start'):
                action = self.start
            elif (message_list[0] == 'stop'):
                action = self.stop
            elif (message_list[0] == 'restart'):
                action = self.restart
            else:
                action = self.status
            if (message_list[0] == 'reload'):
                self.current_client.send(b"Error: Too many argument")
                print('Error: Too many argument')
                return
            del message_list[0]
            if message_list[0] == 'all' and len(message_list) != 1 and action != self.status:
                self.current_client.send(b"Error: Too many argument")
                print('Error: Too many argument')
                return
            if message_list[0] == 'all' and action == self.status:
                self.current_client.send(b"Error: Too many argument")
                print('Error: Too many argument')
                return
            if (message_list[0] == 'all'):
                self.arg_all = True
                message_list = []
                for key in self.program_dict.keys():
                    message_list.append(key)
            self.multipleAction(message_list, action)
            self.arg_all = False

    def receiveMsg(self, client, address):

        buff = 1024
        message = b""
        while message != b"shutdown":
            try:
                message = client.recv(buff)
                self.current_client = client
                if message:
                    print("Received: ", message.decode())
                else:
                    #
                    # A CHECK CAR error n'existe pas
                    #
                    raise error('Client disconnected')
                self.redirectMessage(message.decode())
            except:
                self.closeCo(client)
                return False
        self.redirectMessage("stop all")
        self.sock.close()

    def listen(self):

        try:
            self.sock.listen(5)
            print("Server is listening on port {}" .format(self.port))
        except:
            print("Error: server can't listen on port {}" .format(self.port))
            return
        while True:
            try:
                client_connexion, data_connexion = self.sock.accept()
                print("Accepted a new connexion: {0} on port {1}" .format(data_connexion[0], data_connexion[1]))
                thread = threading.Thread(target=self.receiveMsg, args=(client_connexion, data_connexion))
                thread.start()
            except:
                print('Error: an error occured during acceptation of new client')
                self.sock.close()
                return

    def closeCo(self, client):
        client.close()

    def execute(self, programs_list, program):
        if self.program_dict[program][0] == 'RUNNING':
            if self.arg_all == False:
                self.current_client.send("{}: ERROR (Already started)".format(program).encode('utf-8'))
                print("{}: ERROR (Already started)".format(program))
            return
        child = self.createChild()
        if child != -1:
            if child > 0:
                self.program_dict[program][0] = 'RUNNING'
                self.program_dict[program].append(child)
                self.current_client.send("{}: started".format(program).encode('utf-8'))
                print("{}: started".format(program))
                try:
                    pid, status = os.waitpid(child, 0)
                    self.program_dict[program][0] = 'STOPPED'
                    del self.program_dict[program][1]
                    found_exit = False
                    autorestart = self.conf.parsed_file['programs'][program].get('autorestart', 'unexpected')
                    if autorestart == 'false':
                        found_exit = True
                    if autorestart == 'unexpected':
                        exitcodes = self.conf.parsed_file['programs'][program]['exitcodes'].split(',')
                        for exitcode in exitcodes:
                            if str(status) == str(exitcode):
                                found_exit = True
                    # if found_exit == False:
                    #     self.start(program)

                except OSError as message:
                    print(message)
                return
            if child == 0:
                try:
                    args = programs_list[program]['command'].split(" ")
                    os.execv(args[0], args)
                except OSError as msg:
                    print(msg)
                os._exit(0)

    def createChild(self):
        child_pid = os.fork()
        return child_pid

if __name__ == '__main__':

    new_server = Server('127.0.0.1', 12800).listen()
